### ISSUE TITLE
Privacy Badger failed to initialize

### ISSUE THREAD

**Issue by varuld** [EFForg/privacybadger#2966]
#### What is your browser and browser version?
Firefox_v126.0
#### What is broken and where?
All sites I visit e.g. 'eff.org', 'fsf.org' etc.

**Comment by varuld** [comment]
I.E. this ?
![privacy_bagder_background_1](https://github.com/EFForg/privacybadger/assets/58148529/809f032c-7ce4-4935-b9d3-40755518169f)

**Comment by ghostwords** [comment]
It looks like the console for the eff.org webpage. What we want is the console for Privacy Badger's background page.

**Comment by ghostwords** [comment]
![Screenshot from 2024-05-31 09-25-45](https://github.com/EFForg/privacybadger/assets/794578/179d3c91-7c81-452a-ac02-3b95040a9e53)

**Comment by varuld** [comment]
This should be it
![privacy_bagder_background_2](https://github.com/EFForg/privacybadger/assets/58148529/47ffcba2-3732-418c-acc4-b1ca5ab7532f)

**Comment by ghostwords** [comment]
Just to confirm, this screenshot was taken at the same time when Privacy Badger is showing the exclamation mark on all pages, correct?

**Comment by varuld** [comment]
That is correct

**Comment by ghostwords** [comment]
The screenshot only shows output from running the debugging command, which we don't need to do now. What I am looking for is the error that breaks Privacy Badger when Privacy Badger first starts up.

Could you visit `about:addons` and toggle Privacy Badger off and then back on?

And then could you take and post another screenshot of Privacy Badger's background page? No need to put in any debugging code. I just want to see the errors that come up by themselves.

**Comment by varuld** [comment]
This is the result
![privacy_bagder_background_3](https://github.com/EFForg/privacybadger/assets/58148529/84463f56-d9a8-4fd0-97d0-3a1ea39f8ccd)

**Comment by ghostwords** [comment]
It looks like your Firefox extension storage is failing to return an expected object. I am not sure why we don't see the complete error information ("stack trace"), or why we don't see the storage error. But that's what I think it is.

The good news, sort of, is that this kind of problem was happening before I added the logic to show the "failed to initialize" error message. It's just that before we didn't know that your Privacy Badger was broken, and now we do. So now that we know, we'll try to fix or work around this problem.

**Comment by ghostwords** [comment]
This is related to #2954.

**Comment by varuld** [comment]
> It looks like your Firefox extension storage is failing to return an expected object. I am not sure why we don't see the complete error information ("stack trace"), or why we don't see the storage error. But that's what I think it is.
> 
> The good news, sort of, is that this kind of problem was happening before I added the logic to show the "failed to initialize" error message. It's just that before we didn't know that your Privacy Badger was broken, and now we do. So now that we know, we'll try to fix or work around this problem.

Ping me if you need testing/triage on possible solutions.

**Comment by ghostwords** [comment]
If you use any other extensions, have you noticed anything strange about them? For example, your extension settings get reset to defaults after restarting the browser.

**Comment by ppseprus** [comment]
Having the "same" issue on Chrome, version 125.0.6422.142 (Official Build) (arm64)

But my background,js log is empty, and contradictorily, it says its ready... 🤔  
<img width="983" alt="Screenshot 2024-05-31 at 22 02 06" src="https://github.com/EFForg/privacybadger/assets/1323285/b48f4c56-cbd6-41cf-90b9-ca19c6a01604">

I've just refreshed my browser and before that I hade no issues at all.

No other extension is affected.

**Comment by ppseprus** [comment]
Actually, I tried restarting my browser for the nth time, and now I get this:

```log
Unchecked runtime.lastError: No tab with id: 1509629437.
```

<img width="986" alt="Screenshot 2024-05-31 at 22 06 24" src="https://github.com/EFForg/privacybadger/assets/1323285/33c53efe-06d6-4c35-9433-1335dccfc0ab">

**Comment by ghostwords** [comment]
Hello and thanks for debugging!

>But my background,js log is empty, and contradictorily, it says its ready...

That is strange but good to know.

>Unchecked runtime.lastError: No tab with id: 1509629437.

I think that error is unrelated.

**Comment by ghostwords** [comment]
@ppseprus It looks like a screenshot of the dev tools console for Privacy Badger's popup. Could you try opening the dev tools console for Privacy Badger's background process, next time the red exclamation comes up?

To reach the background process in Chrome, visit `chrome://extensions/`, enable developer mode (in the upper right), and click on the "service worker" link inside Privacy Badger's listing.

**Comment by ppseprus** [comment]
@ghostwords , it opens the same thing. I tried both ways (from extensions, clicking the service worker label AND the popup option in the dropdown.

> next time the red exclamation comes up?
Since I updated the browser about an hour ago, PB stopped working on every single tab. No matter which one I open up. I completely stopped working.

**Comment by ghostwords** [comment]
What do you mean by "stopped working"? Is the popup blank?

Does restarting Privacy Badger help? To restart, visit `chrome://extensions/` and toggle Privacy Badger off and then back on.

**Comment by ppseprus** [comment]
This is what I have on every single tab:
<img width="449" alt="Screenshot 2024-05-31 at 23 12 07" src="https://github.com/EFForg/privacybadger/assets/1323285/780aa3ff-5f8b-4b6c-a21f-1c0626ffc7b3">

> Does restarting Privacy Badger help?

Ahh... okay. I was restarting the browser. Restarting the extension works. Sorry.

**Comment by ghostwords** [comment]
OK thank you!

I'll upload a workaround update shortly. I'll let you know when it goes live so that we see if things got better. Might happen by tomorrow, but maybe Monday.

**Comment by ghostwords** [comment]
Huh, I wonder if restarting the browser can somehow lead to this error. Like, something about extension storage not being ready even though extensions are already running and expect `chrome.storage.local.get()` to return an object, not `undefined`.

**Comment by varuld** [comment]
> If you use any other extensions, have you noticed anything strange about them? For example, your extension settings get reset to defaults after restarting the browser.

I have not. 
It may happen that there happening something of that sort, but I haven't noticed anything.

**Comment by ghostwords** [comment]
My latest theory is that there are two separate issues that people run into:

- (1) Privacy Badger can take longer than ten seconds to finish initializing. This seems to happen on browser startup. We show the "failed to initialize" message on all pages, but we do in fact finish initializing successfully and everything is operational, except for the error badge and the error message in the popup.

- (2) Privacy Badger sometimes fails to read from extension storage during its initialization. We show the "failed to initialize" message on all pages, and Privacy Badger is completely broken at this point. This is probably the more uncommon scenario.

As a workaround, users can restart Privacy Badger by visiting their browser's extensions management page, and toggling Privacy Badger off and then back on. I am not sure this workaround helps when PB fails to read from extension storage though.

We're about to release a partial fix for (2), when PB fails to read from extension storage. Rather than failing completely, Privacy Badger will initialize as if freshly installed, with all settings at defaults. We won't show the new user welcome page though.

We are going to follow this up with a fix for (1) where we deactivate the "failed to initialize" error when PB does in fact finish initializing successfully.

Also for (1), I am going to look into if we can do anything about Privacy Badger taking longer than ten seconds to initialize.

Also for (2), I am also going to look into retrying reading from extension storage, rather than giving up after the first attempt (2).

**Comment by whatinthenameofgod** [comment]
I am getting this exact issue now on Chrome.

**Comment by ghostwords** [comment]
> I am getting this exact issue now on Chrome.

Try restarting Privacy Badger by going to `chrome://extensions` and toggling Privacy Badger off and then back on.

**Comment by sunwolf01** [comment]
Same Problem with:
Firefox_v126.0.1
OS: Windows 10
probably 2nd theory (TypeError: can't convert undefined to object)
restarting of Privacy Badger is not solving the problem, nor restarting firefox - as well not working

**Comment by ghostwords** [comment]
> probably 2nd theory (TypeError: can't convert undefined to object)

Yes, I think so. Then the following earlier posts apply to your problem:

> The good news, sort of, is that this kind of problem was happening before I added the logic to show the "failed to initialize" error message. It's just that before we didn't know that your Privacy Badger was broken, and now we do. So now that we know, we'll try to fix or work around this problem.

> We're about to release a partial fix for (2), when PB fails to read from extension storage. Rather than failing completely, Privacy Badger will initialize as if freshly installed, with all settings at defaults. We won't show the new user welcome page though.

We are waiting on approval from Mozilla, and that unfortunately always seems to take a few days these days.

**Comment by AndyPT** [comment]
For reference, this has also been happening in Edge since last week. 
Microsoft Edge 125.0.2535.67 (Official build) (64-bit) 
Chromium version 125.0.6422.112
Windows 11 (non insider build)

**Comment by willaaa** [comment]
> > I am getting this exact issue now on Chrome.
> 
> Try restarting Privacy Badger by going to `chrome://extensions` and toggling Privacy Badger off and then back on.

This worked for me

**Comment by g-rodigy** [comment]
![image](https://github.com/EFForg/privacybadger/assets/30372087/4dc35e2c-4ef0-4d48-830d-6e75133f7df3)
found this erros

**Comment by ghostwords** [comment]
Hi @g-rodigy , these should be harmless warnings, unrelated to anything in this issue.

**Comment by ghostwords** [comment]
>We're about to release a partial fix for (2), when PB fails to read from extension storage. Rather than failing completely, Privacy Badger will initialize as if freshly installed, with all settings at defaults. We won't show the new user welcome page though.
>
>We are going to follow this up with a fix for (1) where we deactivate the "failed to initialize" error when PB does in fact finish initializing successfully.

These fixes are live as of Privacy Badger version 2024.6.3.

Thanks everybody for your help and patience!

**Comment by ghostwords** [comment]
I'm still looking into followup improvements.

>Also for (1), I am going to look into if we can do anything about Privacy Badger taking longer than ten seconds to initialize.

The initialization delays seem to come from trying to read from ["managed" extension storage](https://github.com/EFForg/privacybadger/blob/master/doc/admin-deployment.md).

The `chrome.storage.managed.get()` call get take a wildly long time to come back when done during browser startup, like ten seconds or longer. I can reproduce this behavior in MV2, but it might have gotten worse in MV3.

We may be able to work around this browser bug by not waiting to read from managed storage. Unfortunately, this means that users' policy-set preferences will not get applied until some time after the websites and Privacy Badger both got going. But if it's between Privacy Badger not being ready to do its job for ten+ seconds on browser startup for all users, and Privacy Badger failing to apply admin policy-set preferences into account until ten+ seconds on browser setup for admin policy users, I think the latter is the less-bad option.

>Also for (2), I am going to look into retrying reading from extension storage, rather than giving up after the first attempt fails.

Apparently, regular extension storage can silently fail and return `undefined` instead of the expect storage when calling `chrome.storage.local.get()` during browser startup. We worked around this by having PB use default settings, which is better than failing completely. We will add limited retrying to reading from extension storage so that if your PB fails to read from storage on startup, it may take a second or two longer to initialize but with your actual local settings read and applied.

**Comment by AndyPT** [comment]
Things seem to be working for Edge also (at least no error message shown).
Thank you!

**Comment by Aenimus** [comment]
I am getting the same issue no matter how many times I restart privacy badger or my browser.

Chrome: Version 125.0.6422.142 (Official Build) (arm64)
Privacy Badger: 2024.6.3

**Comment by ghostwords** [comment]
Hi @Aenimus, could you open a new issue please? We are going to want to debug what's going on, but I don't want to reuse this thread, as the "failed to initialize" problem seems to be fixed for most users.

**Comment by Aenimus** [comment]
Understood and my apologies @ghostwords.
https://github.com/EFForg/privacybadger/issues/2979 for reference.

**Comment by ghostwords** [comment]
These followup improvements described in https://github.com/EFForg/privacybadger/issues/2966#issuecomment-2153316523 are live as of Privacy Badger version 2024.6.14.
